File: codes/go/gpt/complete/task_2/Solution.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
)

func reallocateBed(patient map[string]interface{}, hospitals []interface{}) map[string]interface{} {
	result := map[string]interface{}{
		"patient_id":         patient["id"],
		"assigned_hospital":  nil,
		"assigned_ward":      nil,
		"assigned_bed":       nil,
		"reallocated_patients": []interface{}{},
	}

	isHighRisk := patient["is_post_surgery"].(bool)

	// Step 1: Direct Assignment
	for _, hospitalInterface := range hospitals {
		hospital := hospitalInterface.(map[string]interface{})
		model := hospital["model"].(float64)

		if isHighRisk && model != 4 {
			continue
		}
		if !isHighRisk && model == 4 {
			continue
		}

		for _, wardInterface := range hospital["wards"].([]interface{}) {
			ward := wardInterface.(map[string]interface{})
			if !matchesWard(patient, ward) {
				continue
			}

			availableBeds := ward["available_beds"].([]interface{})
			if len(availableBeds) == 0 {
				continue
			}

			// Assign to the first available bed (lowest number)
			sort.Slice(availableBeds, func(i, j int) bool {
				return availableBeds[i].(float64) < availableBeds[j].(float64)
			})

			result["assigned_hospital"] = hospital["id"]
			result["assigned_ward"] = ward["ward_name"]
			result["assigned_bed"] = availableBeds[0]

			return result
		}
	}

	// Step 2: Reallocation for High-Risk Patients
	if isHighRisk {
		for _, hospitalInterface := range hospitals {
			hospital := hospitalInterface.(map[string]interface{})
			if hospital["model"].(float64) != 4 {
				continue
			}

			for _, wardInterface := range hospital["wards"].([]interface{}) {
				ward := wardInterface.(map[string]interface{})
				if !matchesWard(patient, ward) {
					continue
				}

				currentPatients := ward["current_patients"].([]interface{})
				sort.Slice(currentPatients, func(i, j int) bool {
					p1 := currentPatients[i].(map[string]interface{})
					p2 := currentPatients[j].(map[string]interface{})
					if p1["days_in_hospital"].(float64) == p2["days_in_hospital"].(float64) {
						return p1["bed_number"].(float64) < p2["bed_number"].(float64)
					}
					return p1["days_in_hospital"].(float64) > p2["days_in_hospital"].(float64)
				})

				for _, patientInterface := range currentPatients {
					patientToMove := patientInterface.(map[string]interface{})
					if patientToMove["days_in_hospital"].(float64) <= 3 || patientToMove["non_transferable"].(bool) {
						continue
					}

					newHospital, newWard, newBed := findReallocation(patientToMove, hospitals)
					if newHospital != nil {
						result["assigned_hospital"] = hospital["id"]
						result["assigned_ward"] = ward["ward_name"]
						result["assigned_bed"] = patientToMove["bed_number"]

						reallocatedPatient := map[string]interface{}{
							"patient_id":   patientToMove["id"],
							"new_hospital": newHospital,
							"new_ward":     newWard,
							"new_bed":      newBed,
						}
						result["reallocated_patients"] = append(result["reallocated_patients"].([]interface{}), reallocatedPatient)

						return result
					}
				}
			}
		}
	}

	return result
}

func matchesWard(patient map[string]interface{}, ward map[string]interface{}) bool {
	patientGender := patient["gender"].(string)
	patientAge := patient["age"].(float64)
	patientSpecialRequirements := patient["special_requirements"].([]interface{})

	wardGenderRestriction := ward["gender_restriction"].(string)
	wardAgeRestriction := ward["age_restriction"].([]interface{})
	wardSpecialRequirements := ward["special_requirements"].([]interface{})

	// Gender Restriction
	if wardGenderRestriction == "Male Only" && patientGender != "Male" {
		return false
	}
	if wardGenderRestriction == "Female Only" && patientGender != "Female" {
		return false
	}

	// Age Restriction
	if patientAge < wardAgeRestriction[0].(float64) || patientAge > wardAgeRestriction[1].(float64) {
		return false
	}

	// Special Requirements
	wardSpecialRequirementsMap := make(map[string]bool)
	for _, req := range wardSpecialRequirements {
		wardSpecialRequirementsMap[req.(string)] = true
	}
	for _, req := range patientSpecialRequirements {
		if !wardSpecialRequirementsMap[req.(string)] {
			return false
		}
	}

	return true
}

func findReallocation(patient map[string]interface{}, hospitals []interface{}) (interface{}, interface{}, interface{}) {
	patientGender := patient["gender"].(string)
	patientAge := patient["age"].(float64)
	patientSpecialRequirements := patient["special_requirements"].([]interface{})

	for _, hospitalInterface := range hospitals {
		hospital := hospitalInterface.(map[string]interface{})
		if hospital["model"].(float64) == 4 {
			continue
		}

		for _, wardInterface := range hospital["wards"].([]interface{}) {
			ward := wardInterface.(map[string]interface{})
			if !matchesWard(patient, ward) {
				continue
			}

			availableBeds := ward["available_beds"].([]interface{})
			if len(availableBeds) == 0 {
				continue
			}

			sort.Slice(availableBeds, func(i, j int) bool {
				return availableBeds[i].(float64) < availableBeds[j].(float64)
			})

			return hospital["id"], ward["ward_name"], availableBeds[0]
		}
	}

	return nil, nil, nil
}

func main() {
	var input struct {
		Patient   map[string]interface{} `json:"patient"`
		Hospitals []interface{}          `json:"hospitals"`
	}

	decoder := json.NewDecoder(os.Stdin)
	err := decoder.Decode(&input)
	if err != nil {
		fmt.Println("Error decoding JSON:", err)
		return
	}

	result := reallocateBed(input.Patient, input.Hospitals)

	output, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		fmt.Println("Error encoding JSON:", err)
		return
	}

	fmt.Println(string(output))
}

File: codes/javascript/gpt/complete/task_2/Solution.js
function reallocateBed(patient, hospitals) {
    function matchesWard(patient, ward) {
        const genderMatch = ward.gender_restriction === 'No Restriction' || 
            (patient.gender === 'Male' && ward.gender_restriction === 'Male Only') ||
            (patient.gender === 'Female' && ward.gender_restriction === 'Female Only');
        const ageMatch = patient.age >= ward.age_restriction[0] && patient.age <= ward.age_restriction[1];
        const specialMatch = patient.special_requirements.every(requirement => 
            ward.special_requirements.includes(requirement)
        );
        return genderMatch && ageMatch && specialMatch;
    }

    function findAndAssignBed(patient, hospitals, model4Only) {
        for (const hospital of hospitals) {
            if (model4Only && hospital.model !== 4) continue;
            if (!model4Only && hospital.model === 4) continue;
            for (const ward of hospital.wards) {
                if (matchesWard(patient, ward) && ward.available_beds.length > 0) {
                    const assignedBed = ward.available_beds.sort((a, b) => a - b)[0];
                    return {
                        patient_id: patient.id,
                        assigned_hospital: hospital.id,
                        assigned_ward: ward.ward_name,
                        assigned_bed: assignedBed,
                        reallocated_patients: []
                    };
                }
            }
        }
        return null;
    }

    function findReallocation(patient, hospitals) {
        const candidates = [];
        for (const hospital of hospitals) {
            if (hospital.model !== 4) continue;
            for (const ward of hospital.wards) {
                if (matchesWard(patient, ward)) {
                    for (const currentPatient of ward.current_patients) {
                        if (currentPatient.days_in_hospital > 3 && !currentPatient.non_transferable) {
                            candidates.push({
                                currentPatient,
                                hospital,
                                ward
                            });
                        }
                    }
                }
            }
        }

        candidates.sort((a, b) => {
            if (a.currentPatient.days_in_hospital !== b.currentPatient.days_in_hospital) {
                return b.currentPatient.days_in_hospital - a.currentPatient.days_in_hospital;
            }
            return a.currentPatient.bed_number - b.currentPatient.bed_number;
        });

        for (const candidate of candidates) {
            const { currentPatient, hospital, ward } = candidate;
            for (const targetHospital of hospitals) {
                if (targetHospital.model >= hospital.model) continue;
                for (const targetWard of targetHospital.wards) {
                    if (matchesWard(currentPatient, targetWard) && targetWard.available_beds.length > 0) {
                        const newBed = targetWard.available_beds.sort((a, b) => a - b)[0];
                        return {
                            patient_id: patient.id,
                            assigned_hospital: hospital.id,
                            assigned_ward: ward.ward_name,
                            assigned_bed: currentPatient.bed_number,
                            reallocated_patients: [{
                                patient_id: currentPatient.id,
                                new_hospital: targetHospital.id,
                                new_ward: targetWard.ward_name,
                                new_bed: newBed
                            }]
                        };
                    }
                }
            }
        }
        return null;
    }

    if (patient.is_post_surgery) {
        const directAssignment = findAndAssignBed(patient, hospitals, true);
        if (directAssignment) return directAssignment;

        const reallocation = findReallocation(patient, hospitals);
        if (reallocation) return reallocation;
    } else {
        const directAssignment = findAndAssignBed(patient, hospitals, false);
        if (directAssignment) return directAssignment;
    }

    return {
        patient_id: patient.id,
        assigned_hospital: null,
        assigned_ward: null,
        assigned_bed: null,
        reallocated_patients: []
    };
}

const fs = require('fs');
const input = JSON.parse(fs.readFileSync(0, 'utf-8'));
const result = reallocateBed(input.patient, input.hospitals);
console.log(JSON.stringify(result));